File: src/glasso.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Author: Luigi Augugliaro
! e-mail: luigi.augugliaro@unipa.it
! webpage: http://dssm.unipa.it/augugliaro
!
! version: 1.0.0
! Data: July 27, 2018
!
subroutine glasso(p,S,w,pendiag,nrho,rhoratio,rho,maxR2,maxit,thr,Sgm,Tht,Adj,df,R2,ncomp,Ck,pk,nit,conv,trace)
implicit none
integer :: p,pendiag,nrho,maxit,Adj(p,p,nrho),df(nrho),ncomp(nrho),Ck(p,nrho),pk(p,nrho),nit(nrho),conv,trace
double precision :: S(p,p),w(p,p),rhoratio,rho(nrho),maxR2,thr,Sgm(p,p,nrho),Tht(p,p,nrho),R2(nrho)
! internal variables
integer :: i,j,k
double precision :: Sgm_k(p,p),Tht_k(p,p),maxrho,minrho,ratio,rho_gl(p,p),R2num,R2den
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! computing maxrho            !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
maxrho = 0.d0
do j = 1, p
    do i = j + 1, p
        if(w(1, 1).lt.0.d0) then
            maxrho = max(maxrho, abs(S(i, j)))
        else
            if(w(i, j).gt.0.d0) then
                maxrho = max(maxrho, abs(S(i, j) / w(i, j)))
            end if
        end if
    end do
end do
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! computing R2den             !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
R2den = 0.d0
do j = 1, p
    if(pendiag.eq.1) R2den = R2den + (maxrho * w(j, j))**2
    do i = j + 1, p
        R2den = R2den + S(i, j)**2
    end do
end do
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! computing rho-values        !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
if(rho(1).lt.0.d0) then
    rho(1) = maxrho
    minrho = rhoratio * maxrho
    ratio = exp((log(minrho) - log(maxrho)) / (nrho - 1.d0))
    do k = 2, nrho
        rho(k) = ratio * rho(k - 1)
    end do
end if
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Inizializing Sgm_k and Tht_k            !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Sgm_k = 0.d0
Tht_k = 0.d0
do j = 1, p
    Sgm_k(j, j) = S(j, j)
    if(pendiag.eq.1) then
        if(w(j, j).lt.0.d0) then
            Sgm_k(j, j) = Sgm_k(j, j) + maxrho
        else
            Sgm_k(j, j) = Sgm_k(j, j) + maxrho * w(j, j)
        end if
    end if
    Tht_k(j, j) = 1.d0 / Sgm_k(j, j)
end do
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Starting optimization                  !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
do k = 1, nrho
    if(trace.eq.2) call glasso_trace2_1(k, rho(k))
    rho_gl = rho(k)
    if(w(1, 1).ge.0.d0) rho_gl = rho_gl * w
    call glassosub(p,S,pendiag,rho_gl,maxit,thr,Sgm_k,Tht_k,ncomp(k),Ck(:,k),pk(:,k),nit(k),conv,trace)
    if(trace.eq.1) call glasso_trace1(k, rho(k), nit(k))
    if(conv.ne.0) then
        nrho = k - 1
        exit
    end if
    Sgm(:, :, k) = Sgm_k
    Tht(:, :, k) = Tht_k
    do j = 1, p
        do i = j + 1, p
            if(abs(Tht_k(i, j)).gt.0.d0) then
                df(k) = df(k) + 1
                Adj(i, j, k) = 1
                Adj(j, i, k) = 1
            end if
        end do
    end do
    R2num = 0.d0
    do j = 1, p
        do i = j, p
            R2num = R2num + (S(i, j) - Sgm_k(i, j))**2
        end do
    end do
    R2(k) = 1.d0 - R2num / R2den
    if(R2(k).ge.maxR2) then
        nrho = k
        exit
    end if
end do
end subroutine glasso

File: R/glasso.R
glasso <- function(X, weights, pendiag = FALSE, nrho = 50L, rho.min.ratio, rho, maxR2, maxit = 1.0E+4, thr = 1.0e-04, trace = 0L){
    this.call <- this.call <- match.call()
    # checking 'X'
    if(missing(X)) stop(sQuote("X"), " is missing")
    if(!is.matrix(X)) stop(sQuote("X"), " is not a matrix")
    if(is.null(colnames(X))) {
        vnames <- paste("X", 1:p, sep = "")
        colnames(X) <- vnames
    } else vnames <- colnames(X)
    n <- dim(X)[1]
    p <- dim(X)[2]
    # checking 'weights'
    if(missing(weights)) weights <- matrix(rep(-1, p * p), nrow = p, ncol = p, dimnames = list(vnames, vnames))
    else {
        if(!is.matrix(weights)) stop(sQuote("weights"), " is not a matrix")
        if(dim(weights)[1] != dim(weights)[2]) stop(sQuote("weights"), " is not a square matrix")
        if(!isSymmetric(weights)) stop(sQuote("weights"), " is not a symmetric matrix")
        if(any(weights[weights != 0] < 0)) stop("negative weights are not allowed")
        weights[weights == +Inf] <- .Machine$double.xmax
    }
    # checking 'pendiag'
    if(!is.logical(pendiag)) stop(sQuote("pendiag"), " is not an object of type ", dQuote("logical"))
    if(!is.vector(pendiag)) stop(sQuote("pendiag"), " is not a vector of length ", sQuote(1))
    if(length(pendiag) != 1) stop(sQuote("pendiag"), " is not a vector of length ", sQuote(1))
    # checking 'nrho'
    if(!is.vector(nrho)) stop(sQuote("nrho"), " is not a vector of length ", sQuote(1))
    if(length(nrho) != 1) stop(sQuote("nrho"), " is not a vector of length ", sQuote(1))
    if(abs(as.integer(nrho) - nrho) > 0) stop(sQuote("nrho"), " is not an object of type ", dQuote("integer"))
    if(nrho <= 0) stop(sQuote("nrho"), " is not a positive integer")
    # checking 'rho.min.ratio'
    if(missing(rho.min.ratio)) rho.min.ratio <- ifelse(p < n, 1.0e-4, 1.0e-2)
    else {
        if(!is.vector(rho.min.ratio)) stop(sQuote("rho.min.ratio"), " is not a vector of length ", sQuote(1))
        if(length(rho.min.ratio) != 1) stop(sQuote("rho.min.ratio"), " is not a vector of length ", sQuote(1))
        if(rho.min.ratio < 0 | rho.min.ratio > 1) stop(sQuote("rho.min.ratio"), " does not belong to the interval [0, 1]")
    }
    # checking 'rho'
    if(missing(rho)) rho <- rep(-1, nrho)
    else {
        if(!is.vector(rho)) stop(sQuote("rho"), " is not a vector")
        if(any(rho < 0)) stop(sQuote("rho"), " is not a vector of positive values")
        rho <- sort(rho, decreasing = TRUE)
        nrho <- length(rho)
        rho.min.ratio <- rho[1] / rho[nrho]
    }
    # checking 'maxR2'
    if(missing(maxR2)) maxR2 <- ifelse(p < n, 1, 0.9)
    else {
        if(!is.vector(maxR2)) stop(sQuote("maxR2"), " is not a vector of length ", sQuote(1))
        if(length(maxR2) != 1) stop(sQuote("maxR2"), " is not a vector of length ", sQuote(1))
        if(maxR2 < 0 | maxR2 > 1) stop(sQuote("maxR2"), " does not belong to the interval [0, 1]")
    }
    # checking 'maxit'
    if(!is.vector(maxit)) stop(sQuote("maxit"), " is not a vector of length ", sQuote(1))
    if(length(maxit) != 1) stop(sQuote("maxit"), " is not a vector of length ", sQuote(1))
    if(abs(as.integer(maxit) - maxit) > 0) stop(sQuote("maxit"), " is not an object of type ", dQuote("integer"))
    if(maxit <= 0) stop(sQuote("maxit"), " is not a positive integer")
    # checking 'thr'
    if(!is.vector(thr)) stop(sQuote("thr"), " is not a vector of length ", sQuote(1))
    if(length(thr) != 1) stop(sQuote("thr"), " is not a vector of length ", sQuote(1))
    if(thr < 0 ) stop(sQuote("thr"), " is not a vector of positive values")
    # checking 'trace'
    if(!is.vector(trace)) stop(sQuote("trace"), " is not a vector of length ", sQuote(1))
    if(length(trace) != 1) stop(sQuote("trace"), " is not a vector of length ", sQuote(1))
    if(abs(as.integer(trace) - trace) > 0) stop(sQuote("trace"), " is not an object of type ", dQuote("integer"))
    if(!is.element(trace, c(0L, 1L, 2L))) stop(sQuote("trace"), " can be equal to '0', '1' or '2'")
    # calling glasso.fit
    mx <- colMeans(X)
    S <- crossprod(X) / n - outer(mx, mx)
    out <- glasso.fit(S = S, w = weights, pendiag = pendiag, nrho = nrho,
                        rhoratio = rho.min.ratio, rho = rho, maxR2 = maxR2,
                        maxit = maxit, thr = thr, trace = trace)
    # checking convergence
    if(out$conv != 0) {
        nrho <- out$nrho
        if(nrho == 0) stop("glasso does not congerce with error code ", sQuote(out$conv))
        msg <- paste("glasso does not congerce at nrho =", nrho + 1, " with error code", sQuote(out$conv))
        msg <- paste(msg, ifelse(nrho == 1, "\n  The first solution is reported", paste("\n  The first", nrho, "solutions are reported")))
        warning(msg)
        out$rho <- out$rho[1:nrho]
        out$Sgm <- out$Sgm[, , 1:nrho, drop = FALSE]
        out$Tht <- out$Tht[, , 1:nrho, drop = FALSE]
        out$Adj <- out$Adj[, , 1:nrho, drop = FALSE]
        out$df <- out$df[1:nrho]
        out$R2 <- out$R2[1:nrho]
        out$ncomp <- out$ncomp[1:nrho]
        out$Ck <- out$Ck[, 1:nrho, drop = FALSE]
        out$pk <- out$pk[, 1:nrho, drop = FALSE]
        out$nit <- out$nit[1:nrho]
    }
    if(out$nrho != nrho) {
        nrho <- out$nrho
        out$rho <- out$rho[1:nrho]
        out$Sgm <- out$Sgm[, , 1:nrho, drop = FALSE]
        out$Tht <- out$Tht[, , 1:nrho, drop = FALSE]
        out$Adj <- out$Adj[, , 1:nrho, drop = FALSE]
        out$df <- out$df[1:nrho]
        out$R2 <- out$R2[1:nrho]
        out$ncomp <- out$ncomp[1:nrho]
        out$Ck <- out$Ck[, 1:nrho, drop = FALSE]
        out$pk <- out$pk[, 1:nrho, drop = FALSE]
        out$nit <- out$nit[1:nrho]
    }
    # preparing object with S3 class 'glasso'
    out[[1]] <- X
    out <- c(call = this.call, out)
    out$w[out$w == -1] <- 1
    out$pendiag <- as.logical(out$pendiag)
    attributes(out)$names[c(2, 4, 7)] <- c("X", "weights", "rho.min.ratio")
    class(out) <- "glasso"
    out
}
